unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -352,6 +351,177 @@ public int indexOf(UTF8String v, int start) {
     return -1;
   }
 
+  private enum ByteType {FIRSTBYTE, MIDBYTE, SINGLEBYTECHAR};
+
+  private ByteType checkByteType(byte b) {
+    int firstTwoBits = (b >>> 6) & 0x03;
+    if (firstTwoBits == 3) {
+       return ByteType.FIRSTBYTE;
+     } else if (firstTwoBits == 2) {
+      return ByteType.MIDBYTE;
+    } else {
+      return ByteType.SINGLEBYTECHAR;
+    }
+  }
+
+  /**
+   * Return the first byte position for a given byte which shared the same code point.
+   * @param bytePos any byte within the code point
+   * @return the first byte position of a given code point, throw exception if not a valid UTF8 str
+   */
+  private int firstOfCurrentCodePoint(int bytePos) {
+    while (bytePos >= 0) {
+      ByteType byteType = checkByteType(getByte(bytePos));
+      if (ByteType.FIRSTBYTE == byteType || ByteType.SINGLEBYTECHAR == byteType) {
+        return bytePos;
+      }
+      bytePos--;
+    }
+    throw new RuntimeException("Invalid UTF8 string: " + toString());
+  }
+
+  // Locate to the start position in byte for a given code point
+  private int firstByteIndex(int startByteIndex, int startPointIndex, int codePoint) {
+    int i = startByteIndex; // position in byte
+    int c = startPointIndex; // position in character
+    while (i < numBytes && c < codePoint) {
+      i += numBytesForFirstByte(getByte(i));
+      c += 1;
+    }
+    if (i > numBytes) {
+      throw new StringIndexOutOfBoundsException(codePoint);
```

Comment:
Could we use special value (-1, or numBytes) instead of exception? Exception is expensive.